File: sort/1387_sort_integers_by_the_power_value/1387_sort_integers_by_the_power_value.go
package main

import (
	"fmt"
	"sort"
)

func getKth(lo int, hi int, k int) int {
	power := make(map[int]int)
	arr := make([]int, hi-lo+1)
	for i := lo; i <= hi; i++ {
		arr[i-lo] = i
		getPower(i, power)
	}
	sort.Slice(arr, func(i, j int) bool {
		if power[arr[i]] != power[arr[j]] {
			return power[arr[i]] < power[arr[j]]
		}
		return arr[i] < arr[j]
	})
	return arr[k-1]
}

func getPower(x int, power map[int]int) int {
	ans, tmp := 0, x
	for x != 1 {
		if x&1 == 1 {
			x = 3*x + 1
		} else {
			x /= 2
		}
		ans++
		if val, ok := power[x]; ok {
			ans += val
			break
		}
	}
	power[tmp] = ans
	return ans
}

func main() {
	lo := 12
	hi := 15
	k := 2
	res := getKth(lo, hi, k)
	fmt.Println(res)
}

File: sort/1387_sort_integers_by_the_power_value/1387_sort_integers_by_the_power_value.py

class Solution(object):
    def getKth(self, lo, hi, k):
        self.memo = {}
        def helper(n):
            if n == 1:
                return 0
            if n in self.memo:
                return self.memo[n]
            if n % 2 :
                ans = helper((n*3+1)/2)+2
            else:
                ans = helper(n/2)+1
            self.memo[n] = ans
            return ans
        tmp = sorted([[helper(n),n] for n in range(lo, hi+1)])
        return tmp[k-1][1]


lo, hi, k = 12, 15, 2
res = Solution().getKth(lo, hi, k)
print(res)